src/chat/chatAgentRequestHandler.ts

```diff
@@ -0,0 +1,633 @@
+
+        return { metadata: { command: "", correlationId: correlationId } };
+    };
+
+    async function processToolCalls(
+        sqlTools: { tool: LanguageModelChatTool; parameters: string }[],
+        conversationUri: string,
+        replyText: string,
+        copilotService: CopilotService,
+    ): Promise<GetNextMessageResponse> {
+        if (sqlTools.length === 0) {
+            throw new Error("No tools to process.");
+        }
+
+        let result: GetNextMessageResponse;
+        for (const toolCall of sqlTools) {
+            result = await copilotService.getNextMessage(
+                conversationUri,
+                replyText,
+                toolCall.tool,
+                toolCall.parameters,
+            );
+
+            if (result.messageType === MessageType.Complete) {
+                break;
+            }
+        }
+
+        return result!;
+    }
+
+    function prepareRequestMessages(
+        result: GetNextMessageResponse,
+        context: vscode.ChatContext,
+        referenceTexts: string[],
+    ): vscode.LanguageModelChatMessage[] {
+        // Separate system messages from the requestMessages
+        const systemMessages = result.requestMessages
+            .filter((message: LanguageModelRequestMessage) => message.role === MessageRole.System)
+            .map((message: LanguageModelRequestMessage) =>
+                vscode.LanguageModelChatMessage.Assistant(message.text),
+            );
+
+        const historyMessages = context.history
+            .map((historyItem) => {
+                if ("prompt" in historyItem) {
+                    return vscode.LanguageModelChatMessage.User(historyItem.prompt);
+                } else {
+                    const responseContent = historyItem.response
+                        .map((part) => ("content" in part ? part.content : ""))
+                        .join("");
+                    return responseContent.trim()
+                        ? vscode.LanguageModelChatMessage.Assistant(responseContent)
+                        : undefined;
+                }
+            })
+            .filter((msg): msg is vscode.LanguageModelChatMessage => msg !== undefined);
+
+        // Include the reference messages
+        // TODO: should we cut off the reference message or send a warning if it is too long? (especially without selection)
+        const referenceMessages = referenceTexts
+            ? referenceTexts.map((text) => vscode.LanguageModelChatMessage.Assistant(text))
+            : [];
+
+        //Get the new user messages (non-system messages from requestMessages)
+        const userMessages = result.requestMessages
+            .filter((message: LanguageModelRequestMessage) => message.role !== MessageRole.System)
+            .map((message: LanguageModelRequestMessage) =>
+                vscode.LanguageModelChatMessage.User(message.text),
+            );
+
+        // Combine them in order: system messages, history, then new user messages
```

Comment:
comment missing reference messages

Reply:
sure, made the comment generic to avoid of of sync issues.